src/config/statsd.ts

```diff
@@ -42,17 +44,38 @@ function hrtimer() {
 	};
 }
 
-export const elapsedTimeMetrics = (
+/**
+ * Returns a middleware function which produce duration and requests count metrics
+ * @param path The value of the "path" tag. If not set then the rout path will be used
+ */
+export const elapsedTimeMetrics = (path?: string) => (
```

Comment:
Is there a reason why you would bundle all endpoints under the same 'path'?  Wouldn't you want it to be done individually as to reliably get metrics of each one? 

I'd recommend you use `req.originalUrl` as it shows the whole path including query string if there are any.  If you don't want query string, you can parse it out or use `req.baseUrl + req.path` to get the full path without query.  BaseUrl is useful to get the router mountpoint url.

Reply:
So the reason for overriding path is that we can't use **req.path** as a tag. It can contain path parameters which explode metrics cardinality. (We are currently affected by this problems, we have path tag varliations like `/github/subscriptions/123` and `/github/subscriptions/345` while they should have the same tag, for example: `/github/subscruptions/<installation_id>`


2. The reason to bundle all API requests in the router - is to capture responses in case if it is 401. Hence authentication middleware is common for the router and will be executed before the individual routes middleware. So that's why we need to pass "path" as a parameter, because at that point the request won't have `route` parameter and we can't use raw request path for cardinality reason

3. Regarding this comment https://github.com/atlassian/github-for-jira/pull/789#pullrequestreview-767467777 
That's what I tried first, but it didn't work as I needed.
Unfortunately we can't put this middleware on that level. Because the request won't have `route` parameter. (request will only have it if this middleware function was executed as part of handling some route like:

````
app.get(
		"/github/subscriptions/:installationId",
		elapsedTimeMetrics(),
		csrfProtection,
		getGitHubSubscriptions
	);
```
When the middleware will be called here the req.route.path would be equal to "/github/subscriptions/:installationId"
If I put it to the top level

```
app.use(elapsedTimeMetrics())
```
the req.route will be undefined, and req.path will be "/github/subscriptions/12345" which we can't use as path tag.

So unfortunately we can't use it on the top level